src/main/java/com/unstoppabledomains/resolution/ZNS.java

```diff
@@ -0,0 +1,139 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.unstoppabledomains.exceptions.NSExceptionCode;
+import com.unstoppabledomains.exceptions.NSExceptionParams;
+import com.unstoppabledomains.exceptions.NamingServiceException;
+import com.unstoppabledomains.resolution.contracts.HTTPUtil;
+
+public class ZNS extends NamingService {
+    static final String REGISTRY_ADDRESS = "0x9611c53BE6d1b32058b2747bdeCECed7e1216793"; // eth style zil registry
+                                                                                         // address
+    static final String RECORDS_KEY = "records";
+    private final String provider;
+
+    public ZNS(String blockchainProviderUrl) {
+        this.provider = blockchainProviderUrl;
+    }
+
+    @Override
+    public String namehash(String domain) {
+        // Attaching parent value to the end of the domain to make sure it is a first
+        // element in a resulting array
+        domain = domain + ".0000000000000000000000000000000000000000000000000000000000000000";
+        String[] labels = domain.split("\\.");
+
+        Collections.reverse(Arrays.asList(labels));
+        Optional<String> namehash = Arrays.stream(labels)
+                .reduce((parent, label) -> Utilities.sha256(parent + Utilities.sha256(label), true));
+        return "0x" + namehash.get();
+    }
+
+    public Boolean isSupported(String domain) {
+        String[] split = domain.split("\\.");
+        return (split.length != 0 && split[split.length - 1].equals("zil"));
+    }
+
+    public String addr(String domain, String ticker) throws NamingServiceException {
+        String key = "crypto." + ticker.toUpperCase() + ".address";
+        try {
+            return getRecord(domain, key);
+        } catch(NamingServiceException exception) {
+            if (exception.getCode() == NSExceptionCode.RecordNotFound)
+                throw new NamingServiceException(NSExceptionCode.UnknownCurrency, new NSExceptionParams("d|c", domain, ticker));
+            throw exception;
+        }
+    }
+
+    public String ipfsHash(String domain) throws NamingServiceException {
+        return getRecord(domain, "ipfs.html.value");
+    }
+
+    public String email(String domain) throws NamingServiceException {
+        return getRecord(domain, "whois.email.value");
+    }
+
+    public String owner(String domain) throws NamingServiceException {
+        String[] addresses = getRecordAddresses(domain);
+        if (addresses == null || Utilities.isNull(addresses[0])) {
+            throw new NamingServiceException(NSExceptionCode.UnregisteredDomain, new NSExceptionParams("%d", domain));
+        }
+        return addresses[0];
+    }
+
+    private String getRecord(String domain, String key) throws NamingServiceException {
+        try {
+        JsonObject records = this.getAllRecords(domain);
```

Comment:
@JohnnyJumper Whether it possible to query only one key from the Zilliqa API for `getRecord` method?

Reply:
I can query the full smart contract state (all records for all nodes) or I can query only a portion of the contract state (all records for a single node) I couldn't find a way to query a single record from a single node.

I was using this https://apidocs.zilliqa.com/#getsmartcontractsubstate